homeassistant/components/light/mqtt_json.py

```diff
@@ -0,0 +1,229 @@
+                CONF_FLASH_TIME_SHORT,
+                CONF_FLASH_TIME_LONG
+            )
+        }
+    )])
+
+
+class MqttJson(Light):
+    """MQTT JSON light."""
+
+    # pylint: disable=too-many-arguments,too-many-instance-attributes
+    def __init__(self, hass, name, topic, qos, retain,
+                 optimistic, brightness, rgb, flash_times):
+        """Initialize MQTT JSON light."""
+        self._hass = hass
+        self._name = name
+        self._topic = topic
+        self._qos = qos
+        self._retain = retain
+        self._optimistic = optimistic or topic["state_topic"] is None
+        self._state = False
+        if brightness:
+            self._brightness = 255
+        else:
+            self._brightness = None
+
+        if rgb:
+            self._rgb = [0, 0, 0]
+        else:
+            self._rgb = None
+
+        self._flash_times = flash_times
+
+        def state_received(topic, payload, qos):
+            """A new MQTT message has been received."""
+            values = json.loads(payload)
+
+            if values["state"] == "ON":
+                self._state = True
+            elif values["state"] == "OFF":
+                self._state = False
+
+            if self._rgb is not None:
+                try:
+                    red = int(values["color"]["r"])
+                    green = int(values["color"]["g"])
+                    blue = int(values["color"]["b"])
+
+                    self._rgb = [red, green, blue]
+                except KeyError:
+                    pass
+
+            if self._brightness is not None:
+                try:
+                    self._brightness = int(values["brightness"])
+                except KeyError:
+                    pass
+
+            self.update_ha_state()
+
+        if self._topic["state_topic"] is not None:
+            mqtt.subscribe(self._hass, self._topic["state_topic"],
+                           state_received, self._qos)
+
+    @property
+    def brightness(self):
+        """Return the brightness of this light between 0..255."""
+        return self._brightness
+
+    @property
+    def rgb_color(self):
+        """Return the RGB color value."""
+        return self._rgb
+
+    @property
+    def should_poll(self):
+        """No polling needed for a MQTT light."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the device if any."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._state
+
+    @property
+    def assumed_state(self):
+        """Return true if we do optimistic updates."""
+        return self._optimistic
+
+    def turn_on(self, **kwargs):
+        """Turn the device on."""
+        should_update = False
+
+        message = {"state": "ON"}
+
+        if ATTR_RGB_COLOR in kwargs:
+            message["color"] = {
+                "r": kwargs[ATTR_RGB_COLOR][0],
```

Comment:
Mostly preference, with a little bit of extensibility in mind. I like using `color` as the object name, which doesn't tell you what the array values would be.

I also think it makes the other side look nicer. See https://github.com/corbanmailloux/esp-mqtt-rgb-led/blob/539412d95ed19ea8e0647432e7b199b03bbe2700/mqtt_esp8266_rgb.ino#L208-L212